### PR TITLE
Fix wait condition evaluation issue

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

This fixes/reverts an issue caused by #242. Specifically that issue scheduled condition checks as a task and just ran all tasks to completion. The problem is that we don't re-schedule the condition checks after the tasks that the conditions caused to wake up are completed. So we have effectively reverted that issue to return to evaluating conditions inline, but we added a feature to allow `Workflow` access from inside the conditionals.

💥 **WARNING** This is technically a _history_ backwards incompatible change for anyone that has used 1.1.1 and has relied on wait conditions to be woken up after other coroutines set values. After discussion it was determined that short life and limited use of of the 1.1.1 release allows us to fix this 3-week bug. But this can technically cause non-determinism in those upgrading from 1.1.1 in that it can make a condition properly be re-evaluated a second time in a task where it wasn't before. Users before 1.1.1 are not affected.

### CODE REVIEW

**Comment by cretz** [comment]
This has become a bit more challenging on forever-waiting conditions, moving to draft while I work it out...

**Comment by cretz** [comment]
Updated to basically revert #242, please review

File: src/Temporalio/Worker/WorkflowInstance.cs [hunk 1]
```diff
@@ -651,55 +651,48 @@ private void RunOnce(bool checkConditions)
...
 
-        private Task CheckConditionsAsync()
-        {
-            try
-            {
-                foreach (var source in conditions.Where(t => t.Item1()).Select(t => t.Item2))
+                // Check conditions. It would be nice if we could run this in the task scheduler
+                // because then users could have access to the `Workflow` context in the condition
+                // callback. However, this cannot be done because even just running one task in the
+                // scheduler causes .NET to add more tasks to the scheduler. And you don't want to
```
Comment (Quinn-With-Two-Ns) [inline src/Temporalio/Worker/WorkflowInstance.cs:675]: What other tasks is dotnet adding ?

Comment (cretz) [reply src/Temporalio/Worker/WorkflowInstance.cs:675]: Just looked at stack trace, it seems like this is actually in my stack trace logic at https://github.com/temporalio/sdk-dotnet/blob/1f6676bb9563ddb8eb4606a8df650fb540420fc2/src/Temporalio/Worker/WorkflowInstance.cs#L730 (and just below it for the return-type-based one) which is occurring during stack trace test (stack trace disabled by default in .NET). So I could technically disable stack trace continuations during condition checking, but I think the behavior here in the PR of not pushing the condition checks into the task queue is the best anyways.

**Review by Quinn-With-Two-Ns — APPROVED** [review]

**Pull request merged**